### ISSUE TITLE
Loss of features while converting PMMLPiepeline to pmml

### ISSUE THREAD

**Issue by luckaB** [jpmml/jpmml-sklearn#49]
Hi,
I have fitted a JPMMLPipeline and dumped it to .pkl.z as seen in this code
`clf = RandomForestClassifier(n_estimators = 10)
mapper = DataFrameMapper([
    (data_class_m.columns.values, [ContinuousDomain(), StandardScaler()])
])
pipeline = PMMLPipeline([
   ("mapper", mapper),
    ("estimator", clf)
])
pipeline.fit(data_class_m, target_m)
joblib.dump(pipeline, "p.pkl.z", compress=9)`

and then I converted the pipeline to PMML (I tried both ways - directly from python `sklearn2pmml(pipeline, "pipeline.pmml", with_repr=True)` and from command line `java -jar converter-executable-1.2-SNAPSHOT.jar --pkl-input p.pkl.z --pmml-output pipeline.pmml`)

However, after I loaded the pmml into Java using the JPMML, I found out that there are only 63 feature (InputFields) instead of 1000 features that were used for fitting. Even when I load the p.pkl.z file back to Python, I can see all of the 1000 features. Is it possible, that the converter stores only those features that are actually used in the trained model? Or is there any other reason why I lost so many features during the conversion?

**Comment by vruusmann** [comment]
> Is it possible, that the converter stores only those features that are actually used in the trained model?

Exactly, the final step of conversion is "pruning" the PMML document of all unused features.

> Or is there any other reason why I lost so many features during the conversion?

Your model needs those 63 features to compute a prediction. It has no use for the remaining (1000 - 63) = 937 features, so there is no point in storing them in the PMML document, and loading/preparing their values during model execution.

This "pruning" is a feature, not a bug. In you case, you're achieving >90% reduction in computation work (in the model deployment phase) by not wasting compute cycles on those 937 unused features.

**Comment by vruusmann** [comment]
As a remark, Scikit-Learn (and Apache Spark ML) and PMML conceptualize features a bit differently.

Suppose you have a categorical feature. In Scikit-Learn, if you binarize a **_primary_** categorical feature using the `OneHotEncoder` transform, then you will receive an array of **_secondary_** (pseudo categorical-)features, where each array element represents a category level. For example, a 100-level categorical feature is expanded into a 100-element array of binary indicator features. PMML has native support for categorical features, so there is no need to perform such feature expansion.

So, if your dataset contain categorical features (that are binarized using `LabelEncoder + OneHotEncoder` or `LabelBinarizer`), then the PMML document will again contain less features than Scikit-Learn dataframe, because the converter "reverses" the binarization transform.
